### DISCUSSION TITLE
FileUpload on SettingsPage

### DISCUSSION THREAD

**Discussion by lloricode** [filamentphp/filament#3681] [Help] [answered]
Hi, just want to fill my field with upload image, we have `MediaSettingType` class for specific image handling, this is from old project that we want to add filament panel, we have used `mutateFormDataBeforeFill` but FileUpload not showing uploaded file

```php
<?php

namespace App\Filament\Pages\Settings;

use App\Setting\Types\EmailSettingType;
use App\Setting\Types\MediaSettingType;
use Filament\Pages\SettingsPage;
use Money\Money;

abstract class BaseSettingPage extends SettingsPage
{

    public function mutateFormDataBeforeFill(array $data): array
    {
        foreach ($data as $key => $datum) {
            if ($datum instanceof MediaSettingType) {
                $data[$key] = url($datum);
            } elseif ($datum instanceof EmailSettingType) {
                $data[$key] = (string)$datum;
            } elseif ($datum instanceof Money) {
                $data[$key] = $datum->getAmount() / 100;
            }
        }

        return $data;
    }
```

```php

            FileUpload::make('meta_tag_logo')
                ->nullable(),
```

Hope someone help, thank you

**Comment by danharrin** [comment, marked as answer]
1) File upload fields need an array of files, even if they just handle 1 file at a time. The array needs to have UUID keys.
2) They accept the path of a file, not its URL. The path is relative to the file upload's disk. So `url()` is probably wrong too?

Try `$data[$key] = [\Illuminate\Support\Str::uuid() => PATH TO FILE];`

**Reply by lloricode** [reply]
thank you, will try tomorrow for this, and also how about s3 files using url? since our client site is uses s3 disk on production

**Reply by danharrin** [reply]
We will generate the URL from the path if you use an S3 disk

**Reply by lloricode** [reply]
already tried this, but nothing is working

```php
    public function getFileDirectory(): string
    {
        return \Str::of($this->string) // path without `url()`
            ->replace('/storage', storage_path('app/public'))
            ->before('?v=');
    }
```

```php
        foreach ($data as $key => $datum) {
            if ($datum instanceof MediaSettingType) {
                $data[$key] = [(string) Str::uuid()=> $datum->getFileDirectory()];
```

already check the file path and it is exist,

**Reply by danharrin** [reply]
Can you send an example path generated by `getFileDirectory()`

**Reply by lloricode** [reply]
> Can you send an example path generated by `getFileDirectory()`

![Screenshot_2022-08-30_07-52-11](https://user-images.githubusercontent.com/8251344/187318568-5756317d-8d97-43c9-abbd-6837b579b2dc.png)

**Reply by danharrin** [reply]
The paths need to be relative to your disk.... so probably just the `/image/setting-media` bit if your disk root is `/public/media`?

**Reply by lloricode** [reply]
hi this is my `filesystems.php` content

```php
    'links' => [
        public_path('storage') => storage_path('app/public'),
    ]
```

and it works base on you instruction

```php
    public function getFileDirectory(): string
    {
        return \Str::of($this->string)
            ->replace('/storage', '')
            ->before('?v=');
    }
```

thank you @danharrin

**Reply by lloricode** [reply]
but will check if it is work with s3

**Reply by lloricode** [reply]
> We will generate the URL from the path if you use an S3 disk

about this, i already have a link of file from s3, how can i add this to fileUpload field?

**Reply by lloricode** [reply]
it works thanks you @danharrin
